functions/visualizations.py

```diff
@@ -0,0 +1,113 @@
+import re
+from io import BytesIO
+from typing import Optional
+
+import pandas as pd
+from clustergrammer import Network
+from openpyxl import load_workbook
+from google.cloud import storage
+from cidc_api.models import DownloadableFiles, prism
+
+from .settings import GOOGLE_DATA_BUCKET
+from .util import BackgroundContext, extract_pubsub_data, sqlalchemy_session
+
+
+def vis_preprocessing(event: dict, context: BackgroundContext):
+    with sqlalchemy_session() as session:
+        file_id = extract_pubsub_data(event)
+        file_record: DownloadableFiles = DownloadableFiles.find_by_id(
+            file_id, session=session
+        )
+
+        if not file_record:
+            raise Exception(f"No downloadable file with id {file_id} found.")
+
+        file_blob = _get_file_from_gcs(GOOGLE_DATA_BUCKET, file_record.object_url)
+
+        # Apply the transformations and get derivative data for visualization.
+        for transform_name, transform in _get_transforms().items():
+            vis_json = transform(file_blob, file_record.data_format)
+            if vis_json:
+                # Add the vis config to the file_record
+                setattr(file_record, transform_name, vis_json)
+
+        # Save the derivative data additions to the database.
+        session.commit()
+
+
+def _get_file_from_gcs(bucket_name: str, object_name: str) -> BytesIO:
+    """Download data from GCS to a byte stream and return it."""
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket(bucket_name)
+    blob = bucket.get_blob(object_name)
+    file_str = blob.download_as_string()
+    return BytesIO(bytes(file_str))
+
+
+def _get_transforms() -> dict:
+    """ 
+    Get a list of functions taking an open file and
+    that file's `data_format` as arguments, returning
+    a JSON blob that the frontend will use for visualization.
+    """
+    return {"clustergrammer": _ClustergrammerTransform()}
+
+
+class _ClustergrammerTransform:
+    def __call__(self, data_file, data_format) -> Optional[dict]:
+        """Prepare the data file for visualization in clustergrammer"""
+        fmt = data_format.lower()
+        if not hasattr(self, fmt):
+            return None
+        return getattr(self, fmt)(data_file)
+
+    def npx(self, data_file) -> dict:
+        """Prepare an NPX file for visualization in clustergrammer"""
+        # Load the NPX data into a dataframe.
+        npx_df = _npx_to_dataframe(data_file)
+
+        # TODO: find a better way to handle missing values
+        npx_df.fillna(0, inplace=True)
+
+        # Produce a clustergrammer JSON blob for this dataframe.
+        net = Network()
```

Comment:
nit: can we name it something like `ClusterGramNetwork` or something so it would be obvious that it's a CG specific and not a general "Network" of some kind?